Add a SensitiveBufferAllocator

chrisvest opened the pull request:
Motivation:
We have use cases in PemPrivateKey and sensitive PemValue objects, for buffers that can zero out themselves when closed.
Users may have use cases for sensitive buffers as well.

Modification:
Add a SensitiveBufferAllocator that functions similarly to the unpooled off-heap allocator, except the buffers are equipped with a drop implementation that zero out the memory of the buffer when it is deallocated.
The sensitive buffers otherwise behave exactly the same as any other off-heap buffer.
The allocator is not pooling because we don't anticipate a high allocation rate for sensitive buffers.
The sensitive buffers only come in an off-heap variant, since a GC may freely produce copies of any on-heap memory, eliminating our control of the lifetime of the data.
In the process of adding this, some unnecessary usages of ArcDrop have also been removed to aid testing, and incidentally also reduces overhead.

Result:
It is possible to reliably allocate buffers that zero out their memory when deallocated, thus reducing (but not eliminating) the window for memory scraping attacks.

normanmaurer commented on buffer/src/main/java/io/netty5/buffer/api/unsafe/UnsafeBuffer.java:
```diff
@@ -196,6 +196,7 @@ public Buffer copy(int offset, int length, boolean readOnly) {
             copy.rsize = length;
             copy.roff = 0;
             copy.woff = length;
+            return copy;
```
same comment as above... seems like a bug-fix so should be in another PR .

normanmaurer commented on buffer/src/test/java/io/netty5/buffer/api/tests/BufferTestSupport.java:
```diff
@@ -190,6 +203,12 @@
         // Add 2-way composite buffers of all combinations.
         for (Fixture first : initFixtures) {
             for (Fixture second : initFixtures) {
+                Properties[] properties;
```
```suggestion
                fiinal Properties[] properties;
```

hyperxpro replied:
Typo in `final`.

normanmaurer commented on buffer/src/test/java/io/netty5/buffer/api/tests/SensitiveBufferTest.java:
```diff
@@ -0,0 +1,193 @@
+        }
+    }
+
+    private static class StubManager implements MemoryManager {
```
```suggestion
    private static final class StubManager implements MemoryManager {
```

chrisvest commented on buffer/src/main/java/io/netty5/buffer/api/bytebuffer/NioBuffer.java:
```diff
@@ -184,6 +184,7 @@ public Buffer copy(int offset, int length, boolean readOnly) {
             }
             copy.roff = 0;
             copy.woff = length;
+            return copy;
```
#12262

chrisvest commented:
@normanmaurer I've addressed the comments. The changes in #12262 haven't been reverted from this PR because they are required for the tests to pass. Once #12262 is merged, I can rebase.

chrisvest commented:
Merging this.
